src/core/Akka/IO/TcpConnection.cs

```diff
@@ -46,375 +70,174 @@ namespace Akka.IO
-                        if (!buffered)
-                        {
-                            var writerInfo = new ConnectionInfo(Sender, false, false);
-                            DropWrite(writerInfo, write);
-                        }
-                        else
-                        {
-                            Log.Warning("Received Write command before Register command. " +
-                                        "It will be buffered until Register will be received (buffered write size is {0} bytes)", commandSize);
-                        }
+        private static readonly IOException DroppingWriteBecauseQueueIsFullException =
+            new("Dropping write because queue is full");
 
-                        return true;
-                    default: return false;
-                }
-            };
-        }
+        private int? _partialWriteOffset = null;
 
-        /// <summary>
-        /// Normal connected state.
-        /// </summary>
-        private Receive Connected(ConnectionInfo info)
+        protected TcpConnection(TcpSettings settings, Socket socket, bool pullMode)
         {
-            var handleWrite = HandleWriteMessages(info);
-            return message =>
-            {
-                if (handleWrite(message)) return true;
-                switch (message)
-                {
-                    case SuspendReading _: SuspendReading(); return true;
-                    case ResumeReading _: ResumeReading(); return true;
-                    case SocketReceived _: DoRead(info, null); return true;
-                    case CloseCommand cmd: HandleClose(info, Sender, cmd.Event); return true;
-                    default: return false;
-                }
-            };
-        }
+            Settings = settings;
+            _maxQueuedBytes = settings.WriteCommandsQueueMaxSize; // –1 ⇒ unlimited;
+            _pendingWrites = new Queue<(Write Cmd, IActorRef Sender)>(16);
+            _pullMode = pullMode;
 
-        /// <summary>
-        /// The peer sent EOF first, but we may still want to send 
-        /// </summary>
-        private Receive PeerSentEOF(ConnectionInfo info)
-        {
-            var handleWrite = HandleWriteMessages(info);
-            return message =>
-            {
-                if (handleWrite(message)) return true;
-                var cmd = message as CloseCommand;
-                if (cmd != null)
-                {
-                    HandleClose(info, Sender, cmd.Event);
-                    return true;
-                }
-                if (message is ResumeReading) return true;
-                return false;
-            };
-        }
+            _traceLogging = Settings.TraceLogging;
+            _state = ConnState.Initial(_pendingWrites);
+            Socket = socket ?? throw new ArgumentNullException(nameof(socket));
+            _receiveBuffer = _bufferPool.Rent(settings.MaxFrameSizeBytes);
+            _receiveArgs = new ReadSocketAsyncEventArgs();
```

Comment:
We have exactly 1 `SocketAsyncEventArgs` for reads and 1 for writes. You don't need any more than that for a single TCP conneciton. UDP is a different story.